File: Olympiad/COCI/official/2008/contest5_solutions/avogadro.cpp
#include <cstdio>
#include <stack>
#include <vector>

#define MAX 100000

using namespace std;

int n;
int a[3][MAX];
int erased[MAX];

vector<int> columns[MAX];
int freq[MAX][3];

stack<int> jobs;

int main( void ) {
   scanf( "%d", &n );
   for( int r = 0; r < 3; ++r ) 
      for( int c = 0; c < n; ++c ) {
         scanf( "%d", &a[r][c] ); --a[r][c];

         columns[a[r][c]].push_back( c );
         ++freq[a[r][c]][r];
      }

   for( int i = 0; i < n; ++i ) 
      if( freq[i][1] == 0 || freq[i][2] == 0 ) 
         jobs.push( i );      

   int ret = 0;
   while( !jobs.empty() ) {
      int number = jobs.top();
      jobs.pop();

      for( vector<int>::iterator it = columns[number].begin(); it != columns[number].end(); ++it ) {
         if( erased[*it] ) continue;
         erased[*it] = 1;
         ++ret;

         if( --freq[a[0][*it]][0] == 0 ) jobs.push( a[0][*it] );
         if( --freq[a[1][*it]][1] == 0 ) jobs.push( a[1][*it] );
         if( --freq[a[2][*it]][2] == 0 ) jobs.push( a[2][*it] );
      }
   }

   printf( "%d\n", ret );

   return 0;
}

File: Olympiad/COCI/official/2008/contest5_solutions/avogadro.pas
program avogadro;

const MAX = 100000;

type
   nodeptr = ^node;
   node = record
      data : longint;
      next : nodeptr;
   end;

var
   n : longint;
   a : array[1..3, 1..MAX] of longint;
   erased : array[1..MAX] of boolean;

   columns : array[1..MAX] of nodeptr;
   freq : array[1..MAX,1..3] of longint;

   i, r, c, number, ret : longint;
   it : nodeptr;

   jobsTop : longint;
   jobs : array[1..3*MAX] of longint;

procedure push( x : longint );
begin
   inc( jobsTop );
   jobs[jobsTop] := x;
end;
function pop() : longint;
begin
   pop := jobs[jobsTop];
   dec( jobsTop );
end;

begin
   readln( n );
   for i := 1 to n do begin
      columns[i] := NIL;
      erased[i] := false;
      for r := 1 to 3 do freq[i][r] := 0;
   end;

   for r := 1 to 3 do begin
      for c := 1 to n do begin
         read( a[r,c] );

         new( it );
         it^.data := c;
         it^.next := columns[a[r,c]];
         columns[a[r,c]] := it;

         inc( freq[a[r][c],r] );
      end;
      readln;
   end;

   jobsTop := 0;
   for i := 1 to n do
      if (freq[i,2] = 0) or (freq[i,3] = 0) then
         push( i );

   ret := 0;
   while jobsTop > 0 do begin
      number := pop();

      it := columns[number];
      while it <> NIL do begin
         if erased[it^.data] = false then begin
            erased[it^.data] := true;
            inc( ret );

            for r := 1 to 3 do begin
               dec( freq[a[r,it^.data],r] );
               if freq[a[r,it^.data],r] = 0 then
                  push( a[r,it^.data] );
            end;
         end;
         it := it^.next;
      end;
   end;
   writeln( ret );
end.
